### PR TITLE
feat: Make WebSocket Reader/Writer (#5001)

### LINKED ISSUE #5001: [WS] Make WebSocket Reader/Writer
My use-case for this is piping `WebSocket`<->`Deno.Conn`, e.g.:
```typescript
await Promise.race([Deno.copy(ws, socket), Deno.copy(socket, ws)]);
```
The reason this might be needed is to `websockify` TCP connection (e.g., bridge from `VNC` server to `WebSocket`).

Currently, I have to do the following:

```typescript
const tcpToWebSocket = async () => {
  for await (const data of Deno.toAsyncIterator(socket)) {
    await ws.send(data);
  }
};

const webSocketToTcp = async () => {
  for await (const wsEvent of ws) {
    if (wsEvent instanceof Uint8Array) {
      await socket.write(wsEvent);
    }
  }
};
```

### PR DESCRIPTION
**Pull request by trebler**
Resolves #5001

### CODE REVIEW

**Comment by keroxp** [comment]
@ry Why this merged? WebSocket and TCP are totally different protocol layer. And also this R/W implementation ignores ping/pong/close frame. This abstraction is not good. If use case described in  #5001 is needed, one should do that or use `readFrame`/ `writeFrame` in ws modules.

File: std/ws/mod.ts [hunk 1]
```diff
@@ -327,6 +329,26 @@ class WebSocketImpl implements WebSocket {
...
+  async write(p: Uint8Array): Promise<number> {
+    await this.send(p);
+
+    return p.byteLength;
+  }
+
+  async read(p: Uint8Array): Promise<number | null> {
+    for await (const ev of this.receive()) {
+      if (ev instanceof Uint8Array) {
+        return copyBytes(p, ev);
```
Comment (keroxp) [inline std/ws/mod.ts:341]: `copyBytes` ignores overflowed data of src. It seems to discard data in frames, doesn't it?.

**Comment by trebler** [comment]
> And also this R/W implementation ignores ping/pong/close frame.

@keroxp, the purpose was to only handle data frames / messages. `ping`/`pong`/`close` are "technical" frames. E.g., if one wanted to pipe all WS messages to/from a file, DB, TCP socket, one would want to only read/write "useful" information (being messages).

File: std/ws/mod.ts (hunk 1 above)
Comment (trebler) [reply std/ws/mod.ts:341]: I think you are right. Would something like this make sense instead?

```typescript
private readBacklog: Uint8Array = new Uint8Array()

  async read(p: Uint8Array): Promise<number | null> {
    for await (let ev of this) {
      if (ev instanceof Uint8Array) {
        if (this.readBacklog.byteLength) {
          const copiedFromBacklog = copyBytes(this.readBacklog, p);
          this.readBacklog = this.readBacklog.subarray(copiedFromBacklog);
          
          return copiedFromBacklog + copyBytes(ev, p, copiedFromBacklog);
        }

        const copied = copyBytes(ev, p);
        this.readBacklog = ev.subarray(copied);

        return copied;
      }

      if (typeof ev === "string") {
        return copyBytes(encode(ev), p);
      }

      // if (typeof ev === "string") {
    }

    return null;
  }
```

**Comment by ry** [comment]
> Why this merged? 

@keroxp Negligence on my part. Now that you raise the point, I agree that Reader/Writer isn't the right abstraction for a frame oriented protocol. Let's revert it.

**Pull request merged**